### ISSUE TITLE
fluentd-gcp log collection not working on GCI masters since Monday

### ISSUE THREAD

**Issue by a-robinson** [kubernetes/kubernetes#26379]
Will start investigating shortly

cc @krousey

**Comment by a-robinson** [comment]
@andyzheng0831

**Comment by a-robinson** [comment]
Also, @andyzheng0831, there seems to be a correlation between me SSHing into a GCI master that's being deployed and the cluster creation subsequently failing to complete. It's happened 2 times now...

**Comment by andyzheng0831** [comment]
I did sshing while master was under initialization, but did not see the
problem. BTW, i am ooo and not be able to investigate it until next Tue. My
auto email reply does not work?

**Comment by a-robinson** [comment]
Not for github cc's. Sorry for bothering you!

**Comment by a-robinson** [comment]
It keeps getting blocked on the tarball corruption... Opening a separate issue for that part

**Comment by a-robinson** [comment]
Ok, finally have logs from a failure (it's too bad logs don't get saved when the logging daemon isn't working...):

```
2016-05-26 21:31:11 +0000 [error]: unexpected error error="Could not load the default credentials. Browse to\nhttps://developers.google.com/accounts/docs/application-default-cred
entials\nfor more information\n"
```

That's a little strange though, considering that the VM does have the appropriate scope. Continuing to investigate.

**Comment by a-robinson** [comment]
The metadata endpoint isn't resolvable from within the container. The DNS must be messed up in some way. We've had this happen before but fixed the configuration to unbreak it. I wonder what retriggered it when switching to GCI.

Bumping to P0 because this is blocking investigation of at least one flake.

**Comment by a-robinson** [comment]
cc @ArtfulCoder @bprashanth - the DNS policy for the pod is still set to "Default", so it's not the exact same issue as before. The resolv.conf in the container is

```
search google.internal c.gke-test-us-central1-f-0.internal
nameserver 169.254.169.254
options ndots:5
```

**Comment by a-robinson** [comment]
This affects k8s-on-gce as well

**Comment by bprashanth** [comment]
that resolv.conf looks OK, does nslookup metadata just fail? docker exec in the other master components works?

**Comment by adityakali** [comment]
@kubernetes/goog-image

**Comment by a-robinson** [comment]
I can't actually verify whether nslookup fails, because I can't even apt-get install dnsutils due to not being able to resolve archive.ubuntu.com :/

`ping google.com` doesn't work either. DNS is just completely busted inside the fluentd container.

Hitting the metadata server at its IP address works fine:
```shell
$ curl 169.254.169.254
0.1/
computeMetadata/
```

**Comment by lavalamp** [comment]
@wonderfly @andyzheng0831 since this is probably GCI related.

**Comment by lavalamp** [comment]
@girishkalele since you were talking to me about something that sounds similar.

**Comment by bprashanth** [comment]
> I can't actually verify whether nslookup fails, because I can't even apt-get install dnsutils due to not being able to resolve archive.ubuntu.com :/

But you can docker exec into etcd and nslookup say? i think nslookup is already present in etcd since it's busybox based

**Comment by krousey** [comment]
@a-robinson nslookup works fine on my GCE master. Are you sure this affects k8s on GCE as well?

**Comment by a-robinson** [comment]
Yes, but only in the fluentd container.

@bprashanth - it resolves just fine from within etcd, but that uses host networking.

**Comment by a-robinson** [comment]
fluentd-es's resolv.conf on a master:
```
search kube-system.svc.cluster.local svc.cluster.local cluster.local google.internal c.k8s-jkns-e2e-gce.internal
nameserver 10.0.0.10
options ndots:5
```

etcd's / the host's:
```
# This file is managed by systemd-resolved(8). Do not edit.
#
# Third party programs must not access this file directly, but
# only through the symlink at /etc/resolv.conf. To manage
# resolv.conf(5) in a different way, replace the symlink by a
# static file or a different symlink.
nameserver 169.254.169.254
search google.internal c.k8s-jkns-e2e-gce.internal
```

**Comment by a-robinson** [comment]
Presumably the problem is that the kube-proxy isn't running on the master, so it can't reach the nameserver at 10.0.0.10. I'm not sure why this broke on Monday, though.

**Comment by a-robinson** [comment]
Ok, ignore that case, since that's just a matter of fluentd-es being brought up on the master when it shouldn't be, without dnsPolicy set to Default.

But on the GKE master (and presumably on a gce master if you configure it to use fluentd-gcp), you get this resolv.conf, yet resolution still doesn't work:
```
# cat /etc/resolv.conf 
search google.internal c.gke-test-us-central1-f-0.internal
nameserver 169.254.169.254
options ndots:5
# nslookup metadata
Server:    169.254.169.254
Address 1: 169.254.169.254

nslookup: can't resolve 'metadata'
# nslookup google.com
Server:    169.254.169.254
Address 1: 169.254.169.254

nslookup: can't resolve 'google.com'
```

So it looks like the GCE DNS server isn't reachable from within the container

**Comment by a-robinson** [comment]
```
# ping 169.254.169.254
PING 169.254.169.254 (169.254.169.254): 56 data bytes
^C
--- 169.254.169.254 ping statistics ---
7 packets transmitted, 0 packets received, 100% packet loss
# ifconfig
eth0      Link encap:Ethernet  HWaddr 02:42:AC:11:00:03  
          inet addr:172.17.0.3  Bcast:0.0.0.0  Mask:255.255.0.0
          inet6 addr: fe80::42:acff:fe11:3/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1460  Metric:1
          RX packets:529 errors:0 dropped:0 overruns:0 frame:0
          TX packets:83 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:22285 (21.7 KiB)  TX bytes:3798 (3.7 KiB)
lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:44 errors:0 dropped:0 overruns:0 frame:0
          TX packets:44 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:4088 (3.9 KiB)  TX bytes:4088 (3.9 KiB)
```

**Comment by a-robinson** [comment]
```
# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
^C
--- 8.8.8.8 ping statistics ---
8 packets transmitted, 0 packets received, 100% packet loss
```

**Comment by a-robinson** [comment]
On the host, it's clear no packets are making it through the docker bridge:
```
# ifconfig
cbr0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1460
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 0.0.0.0
        inet6 fe80::b4a9:a2ff:fef2:2ce5  prefixlen 64  scopeid 0x20<link>
        ether 76:04:6c:46:ba:05  txqueuelen 0  (Ethernet)
        RX packets 831  bytes 23892 (23.3 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 12  bytes 738 (738.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
docker0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 0.0.0.0
        ether 02:42:f5:34:c0:a1  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1460
        inet 10.240.0.17  netmask 255.255.255.255  broadcast 10.240.0.17
        inet6 fe80::4001:aff:fef0:11  prefixlen 64  scopeid 0x20<link>
        ether 42:01:0a:f0:00:11  txqueuelen 1000  (Ethernet)
        RX packets 112254  bytes 573292049 (546.7 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 91056  bytes 54518547 (51.9 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 0  (Local Loopback)
        RX packets 318046  bytes 211482829 (201.6 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 318046  bytes 211482829 (201.6 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
vetha02b25e: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1460
        inet6 fe80::7404:6cff:fe46:ba05  prefixlen 64  scopeid 0x20<link>
        ether 76:04:6c:46:ba:05  txqueuelen 0  (Ethernet)
        RX packets 742  bytes 31476 (30.7 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 87  bytes 3855 (3.7 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
vetha57c42e: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1460
        inet6 fe80::b487:c7ff:fedd:2fdd  prefixlen 64  scopeid 0x20<link>
        ether b6:87:c7:dd:2f:dd  txqueuelen 0  (Ethernet)
        RX packets 89  bytes 4050 (3.9 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 736  bytes 30979 (30.2 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

**Comment by bprashanth** [comment]
there's probably a bug with running non-host networked pods on the master. your cbr0 has some 172 address. I'm guessing this doesn't match the podcidr on your master. I think there's an implicit assumption that master components run with host networking. With kubenet (network plugin) we should be creating cbr0 with the right cidr and not even requiring a docker restart, maybe it's still running some old code and the kubelet just didn't restart docker? 172 looks like the defult doker bridge range.  
@freehan

**Comment by Amey-D** [comment]
On the same instance as @a-robinson 's:

# docker run busybox ping -c3 www.google.com
ping: bad address 'www.google.com'

# docker run --net=host busybox ping -c3 www.google.com
PING www.google.com (173.194.195.105): 56 data bytes
64 bytes from 173.194.195.105: seq=0 ttl=55 time=0.638 ms
64 bytes from 173.194.195.105: seq=1 ttl=55 time=0.473 ms
64 bytes from 173.194.195.105: seq=2 ttl=55 time=0.524 ms

 ps -efl | grep 'docker daemon'
4 S root      1041   158  0  80   0 - 300009 futex_ May19 ?       01:31:13 /usr/bin/docker daemon -s overlay --host=fd:// -p /var/run/docker.pid --bridge=cbr0 --iptables=false --ip-masq=false --debug

# iptables -S
-P INPUT DROP
-P FORWARD DROP
-P OUTPUT DROP
-N DOCKER
-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -p icmp -j ACCEPT
-A INPUT -p tcp -m tcp --dport 22 -j ACCEPT
-A INPUT -p tcp -j ACCEPT
-A INPUT -p udp -j ACCEPT
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A FORWARD -p tcp -j ACCEPT
-A FORWARD -p udp -j ACCEPT
-A OUTPUT -m state --state NEW,RELATED,ESTABLISHED -j ACCEPT
-A OUTPUT -o lo -j ACCEPT

Looks like docker is using cbr0, but the iptables rules are set up for docker0 device?

(Side note: n a non-k8s GCI instance:
# docker run busybox ping -c3 www.google.com
PING www.google.com (74.125.141.99): 56 data bytes
64 bytes from 74.125.141.99: seq=0 ttl=54 time=0.725 ms
64 bytes from 74.125.141.99: seq=1 ttl=54 time=0.505 ms
64 bytes from 74.125.141.99: seq=2 ttl=54 time=0.453 ms
--- www.google.com ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.453/0.561/0.725 ms
)

**Comment by Amey-D** [comment]
I manually added some iptables rules and tweaked docker0 on the master instance:

    # iptables -A FORWARD -o cbr0 -j DOCKER
    # iptables -A FORWARD -o cbr0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
    # iptables -A FORWARD -i cbr0 ! -o cbr0 -j ACCEPT
    # iptables -A FORWARD -i cbr0 -o cbr0 -j ACCEPT
    # ip addr del 172.17.0.1/16 dev docker0
    # docker run busybox ping -c3 www.google.com
    PING www.google.com (209.85.200.106): 56 data bytes
    64 bytes from 209.85.200.106: seq=0 ttl=52 time=0.846 ms
    64 bytes from 209.85.200.106: seq=1 ttl=52 time=0.564 ms
    64 bytes from 209.85.200.106: seq=2 ttl=52 time=0.502 ms

    --- www.google.com ping statistics ---
    3 packets transmitted, 3 packets received, 0% packet loss
    round-trip min/avg/max = 0.502/0.637/0.846 ms

After these changes the fluentd container also has network access:

    # docker exec 30b774d9cdf3 ping -c3 www.google.com
    PING www.google.com (74.125.193.147) 56(84) bytes of data.
    64 bytes from ig-in-f147.1e100.net (74.125.193.147): icmp_seq=1 ttl=56 time=0.867 ms
    64 bytes from ig-in-f147.1e100.net (74.125.193.147): icmp_seq=2 ttl=56 time=0.711 ms
    64 bytes from ig-in-f147.1e100.net (74.125.193.147): icmp_seq=3 ttl=56 time=0.699 ms

    --- www.google.com ping statistics ---
    3 packets transmitted, 3 received, 0% packet loss, time 2001ms
    rtt min/avg/max/mdev = 0.699/0.759/0.867/0.076 ms

**Comment by bprashanth** [comment]
I suspect gci trusty uses some other configuration of standalone kubelet on the master, that doesn't understand how to invoke kubenet. If you can gather kubelet logs it would help.

**Comment by a-robinson** [comment]
Would the systemd unit automatically restart the default configured docker when the kubelet kills it, perhaps preventing it from being started with the correct flags?

**Comment by a-robinson** [comment]
(please pardon my currently lacking familiarity with systemd if that's a terrible question)

**Comment by a-robinson** [comment]
Because it looks like the kubelet tried to do the right thing:
```
May 27 19:56:12 gke-310872969129-1d77833365fc92ce332c kubelet[849]: I0527 19:56:May 27 19:56:12 kubelet[849]: I0527 19:56:12.453165     849 container_bridge.go:157] MASQUERADE rule doesn't exist, recreate it (with nonMas
queradeCIDR 10.0.0.0/8)
May 27 19:56:12 kubelet[849]: I0527 19:56:12.454121     849 container_bridge.go:77] CBR0 doesn't exist, attempting to create it with range:
172.17.0.1/16
May 27 19:56:12 kubelet[849]: I0527 19:56:12.522739     849 container_bridge.go:67] Recreated cbr0 and restarted docker
May 27 19:56:12 kubelet[849]: E0527 19:56:12.572563     849 generic.go:197] GenericPLEG: Unable to retrieve pods: An error occurred trying t
o connect: Get http://%2Fvar%2Frun%2Fdocker.sock/containers/json?all=1&limit=0: EOF
May 27 19:56:12 kubelet[849]: W0527 19:56:12.574192     849 image_manager.go:135] [ImageManager] Failed to monitor images: An error occurred
 trying to connect: Get http://%2Fvar%2Frun%2Fdocker.sock/images/json: EOF
May 27 19:56:12 kubelet[849]: I0527 19:56:12.819891     849 factory.go:217] registering Docker factory
```

**Comment by adityakali** [comment]
killing docker (using kill / pkill) will not restart it. If kubelet does "systemctl restart docker", it should come back with any flag changes.

**Comment by bprashanth** [comment]
The only puzzling point is: wth is `172.17.0.1/16`?

the master kubelet should be created with "-reconcile-cidr=false --pod-cidr=10-dot --configure-cbr0=true", and it should apply that cidr to cbr0. I'm guessing that's being piped wrong (or are we really piping 172.17?). If that were to work, your fluent will get an ip from that /30 pod-cidr, everythign else will run with host networking. We can't rely on nodecontroller to allocate a cidr because of the chicken and egg problem. 

That pod-cidr needs a route if we want to ping cross host. I'm guessing we don't for fluent, though. 

@adityakali as a sidenote about restarting docker, can you verify these assumptions weren't overlooked: 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container_bridge.go#L46. If we're running both the healthmonitor and a docker unit file, they'll probably fight when it comes to restarting docker. unless we started the kubelet with --babysit=false we'll pkill docker assuming it gets restarted.

**Comment by krousey** [comment]
Any update on this?

**Comment by adityakali** [comment]
Regarding restarting docker, I see that kubelet has --babysit-daemons=true on both master and nodes (when on GCI). But I do see docker up and running with command-line flag changes. So I don't think docker restart is an issue.

**Comment by bprashanth** [comment]
Yeah the question is what restarts docker when the kubelet does the pkill? ideally we would want a single thing responsible for restarting it, if we have a systemd unit file and a health-checker daemon running there's always room for trickery.  I see both /etc/systemd/system/kube-docker-monitor.service and /usr/lib/systemd/system/docker.service, is this intentional?

**Comment by andyzheng0831** [comment]
It is intentional. kube-docker-monitor.service is for health monitoring,
just like the supervisor stuff in ContainerVM. docker.service is the real
systemd config for docker. kube-docker-monitor.service does not restart
docker, it simply kills docker, and then docker.service restarts it. I
don't see conflict here.

**Comment by bprashanth** [comment]
ah, sorry I didn't actually look into what it was doing. It looks like it was indeed just doing a docker ps and pkill docker, leaving the actual restart up to systemd.

**Comment by andyzheng0831** [comment]
I am debugging it and I guess it may be related with kubenet setting in GCI config. I am make a patch and test it.

**Comment by dchen1107** [comment]
re: https://github.com/kubernetes/kubernetes/issues/26379#issuecomment-222268962

Since we switched to kubernet with 1.3 release, why Kubelet still config cbr0, and restart docker? cc/ @freehan

**Comment by bprashanth** [comment]
Hmm, while deploying kubenet might fix it, configure-cbr0 should still work just like it did without kubenet (or did we consciously break that?).

 Kubenet requires the pod cidr, and there's a mismatch on the master. The nodecontroller assigned cidr is different from the --pod-cidr argument, because we need to start the nodecontroller.

**Comment by freehan** [comment]
Is kubelet running with `--network-plugin=kubenet`?

**Comment by bprashanth** [comment]
not on the master

**Comment by andyzheng0831** [comment]
See https://github.com/kubernetes/kubernetes/issues/26379#issuecomment-223163421

**Comment by andyzheng0831** [comment]
Now I am pretty sure the kubenet config added by PR #18795 is missing in GCI config. We did not catch this because Jenkins on GCI has not reported an error. I will add kubenet support shortly

**Comment by andyzheng0831** [comment]
@dcbw I am planning to add CNI support in GCI (https://github.com/kubernetes/kubernetes/tree/master/cluster/gce/gci). The dir /opt in GCI image is read-only. Would you please make the CNIDir configurable, which currently is hard-coded to "/opt/cni/bin" (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/kubenet/kubenet_linux.go#l54)

**Comment by andyzheng0831** [comment]
I had an offline discussion with @freehan and he will make a PR for the CNI change I mentioned above. We are not 100% sure the missed kubenet config is the root cause of this issue, but we have to first fix it and then continue debug.

**Comment by bprashanth** [comment]
We must either make --configure-cbr0 work without --network-plugin like it did before, or have it explode. Silently doing the wrong this is bad. IMO we should do the former, since people will also inevitably use it with --network-plugin=!kubenet.

**Comment by bprashanth** [comment]
It might also make sense to ditch the kubenet dependency alltogether on the master. All the imporatant pods don't need it because they use host networking, except fluent.

**Comment by andyzheng0831** [comment]
The current status: I am making a PR to add the network plug-in support in GCI.

**Comment by andyzheng0831** [comment]
Status update:
@freehan and I debugged further and found that deleting docker0 solves the access problem. I copy the comment by @freehan below:

"Deleting the docker0 bridge solves the problem. I never saw any problem with coexistence of cbr0 and docker0. But in this case, the cidr range assigned to cbr0 happens to be the default cidr range of docker0. It is unclear why this causes dns request failure from pod."

This also explains why the issue does not happen in k8s, because the pod-cird is much different.

I made a change and now am verifying it.

**Comment by andyzheng0831** [comment]
There are two issues:
(1) GKE master on GCI has a broken fluentd-gcp pod which cannot access the external addresses;
(2) "ping" from a pod running on GCI cluster does not work.

The issue (1) is only for GKE master and only in the master branch. It does not happen in k8s or 1.2 branch. The issue (2) affects both GKE and k8s, and I think it is also in 1.2 branch.

@a-robinson I would suggest to log a separate issue for (2)

**Comment by andyzheng0831** [comment]
PR #27016 was merged, which fixes the fluend-gcp pod issue in GCI. I will close this issue. For the "ping" issue, we already have an internal bug for it. Please feel free to log a github issue if needed.

**Comment by a-robinson** [comment]
Thanks @andyzheng0831!
